### DISCUSSION TITLE
Taro 3.6 版本支持 React Native 0.70

### DISCUSSION THREAD

**Discussion by zhiqingchen** [NervJS/taro#12468] [RFC]
- 提案时间: (2022-09-09)
- 影响版本: (3.x)

## 概述

将默认的 RN 版本切换至 0.70，expo 版本切换至 47。
参考[React Native 0.69 版本支持](https://github.com/NervJS/taro/discussions/12133)

## 动机

[React Native 0.70 版本已于 2022-9-5 正式发布](https://reactnative.dev/blog/2022/09/05/version-070)。0.70 版本 [在 Android 将默认使用 Hermes 引擎](https://reactnative.dev/blog/2022/07/08/hermes-as-the-default)，Hermes 也带来了RN性能的较大提升，特别是启动场景。
Taro 将与 RN 社区保持同步，将默认的 RN 版本设置为 0.70。

## 使用案例

将 Taro 升级至最新版，将原生代码根据相关pr进行变更。

### 依赖变更
```
    "@react-native-community/clipboard": "~1.5.1",
    "@react-native-community/cameraroll": "~4.1.2",
```
替换为：
```
    "@react-native-camera-roll/camera-roll": "~5.0.4",
    "@react-native-clipboard/clipboard": "~1.11.1",
```
### 支持使用 react-native 命令进行打包

> 为方便后续升级以及支持 Hermes，优先使用 react-native 命令进行打包。同时保留 yarn build:rn 的方式。
>
> 变更后，taro rn 的编译将于 rn 过程完全一致，更方便开发者进行定制。

  0. 增加入口文件 `index.js` 内容为空。
  1. 增加 `metro.config.js` 配置，从 rn-support 包中引入 taro 的默认配置。
  2. `yarn start` 命令改为 `react-native start`
  3. 取消 iOS 的 skip bundle
  4. 还原 Android 中 `android/app/build.gradle` 的 `apply from: "../../node_modules/react-native/react.gradle"`
  5. 打包脚本中删除 `yarn build:rn` 相关内容

### Taro 主仓库升级至 0.70

升级内容：

0. 将不兼容的RN依赖库进行升级，并验证其功能可用性。
6. 将 `@types/react-native` 及 `react-native` 进行版本更新。
7. 将 metro、@react-native-community/cli-plugin-metro 相关依赖与 0.70 使用的版本进行同步。
8. 将默认安装的RN版本进行更新 https://github.com/NervJS/taro/blob/next/packages/taro-cli/src/presets/platforms/rn.ts#L33

### Taro Native Shell 升级至 0.70

分支：https://github.com/NervJS/taro-native-shell/tree/0.70.0

参考 [upgrade-helper](https://react-native-community.github.io/upgrade-helper/?from=0.69.5&to=0.70.0)

### Taro Project Template 升级至 0.70

0. 更新代码同 Taro Native Shell 升级至 0.70。

### Taro Playground 升级至 0.70

分支：

https://github.com/wuba/taro-playground/tree/releases/v1.8.0

升级内容：

0. 版本号改为 v1.8.0。
0. 更新代码同 Taro Native Shell 升级至 0.70。

## 缺陷

- 0.70 版本发布时间较短，及时跟进 0.70 问题。
- IPHONEOS_DEPLOYMENT_TARGET 由 12.4 改为 13.0。 需要向下兼容的开发者，需要保留在低版本。
- 默认 js 引擎为 Hermes，debug工具为 flipper，如不需要可自行切换。

## 替代选择

N/A

## 适配策略

- Taro 主工程默认使用 0.70，保证旧版本 RN 仍正常使用新版本的 Taro。
- 有些包（如 metro）可能无法向下兼容，则需要用户使用 resolution 进行版本锁定，需要在版本更新时说明。

### Library not loaded: @rpath/hermes.framework/hermes on iOS

参考 https://github.com/facebook/react-native/issues/34608  需要使用 2.7.5 的 ruby，重新安装 cocoapods 及进行 pod install

### iOS release build crash

参考 https://github.com/expo/expo/issues/19179#issuecomment-1252370945

**Comment by zhiqingchen** [comment]
## Taro Native Shell 升级至 0.70 测试

对应分支

https://github.com/NervJS/taro-native-shell/tree/0.67.0

下载现有包进行验证

https://github.com/NervJS/taro-native-shell/releases/tag/v0.70.0-3

修改代码后

```
# 10 为自增数字
git tag v0.70.0-3
git push origin v0.70.0-3
```

查看打包是否正常进行

https://github.com/NervJS/taro-native-shell/actions

**Comment by zhiqingchen** [comment]
## Taro 主仓库升级至 0.70

对应分支

https://github.com/NervJS/taro/tree/feat/react-native

pr
https://github.com/NervJS/taro/pull/12471

包安装

3.5.5-alpha.0

### 测试要点

无

**Comment by zhiqingchen** [comment]
## Taro Playground 升级至 0.70 测试

对应分支

https://github.com/wuba/taro-playground/tree/releases/v1.8.0

包下载
https://github.com/wuba/taro-playground/releases/tag/v1.8.0-beta.19

**Comment by zhiqingchen** [comment]
## Taro Project Template 升级至 0.70 测试

对应分支

https://github.com/zhiqingchen/taro-project-templates/tree/v3.6

```
# 全局安装beta版本cli
yarn global add @tarojs/cli@canary

# 通过模板进行初始
taro init

# 初始化后安装依赖
yarn upgradePeerdeps
```

1. 修改 assemble_ios_debug.yml 配置
   1. APP_ID: ${{secrets.APP_ID}}
   2. TEAM_ID: ${{secrets.TEAM_ID}}
   3. PROVISIONING_PROFILE_SPECIFIER: ${{secrets.DEBUG_PROVISIONING_PROFILE_SPECIFIER}}
   4. CODE_SIGN_IDENTITY: Apple Development
2. 修改 assemble_ios_release.yml 配置
   1. APP_ID: ${{secrets.APP_ID}}
   2. TEAM_ID: ${{secrets.TEAM_ID}}
   3. PROVISIONING_PROFILE_SPECIFIER: ${{secrets.RELEASE_PROVISIONING_PROFILE_SPECIFIER}}
   4. CODE_SIGN_IDENTITY: Apple Distribution

```
# 将源设置为 shell 工程
git remote add origin git@github.com:NervJS/taro-native-shell.git

# 10 为自增数字
git tag v0.70.0-beta.0
git push origin v0.70.0-beta.0
```

查看打包是否正常进行

https://github.com/NervJS/taro-native-shell/actions

参考分支

https://github.com/NervJS/taro-native-shell/tree/test/template-0.70-0

**Comment by ShaoGongBra** [comment]
taro build:rn 能把代码打包成二进制文件吗

**Comment by zhiqingchen** [comment]
目前不能，后续计划把build交给rn原逻辑。这样开启hermes后，就能编成二进制文件

**Comment by ShaoGongBra** [comment]
47发布bete版本了，好多个expo组件也更新了，我都更新了，ios死循环的问题还在

**Reply by zhiqingchen** [reply]
https://github.com/wuba/taro-playground/tree/releases/v1.8.0
该分支已无死循环问题，可对比看看，是否哪个包版本有误

**Comment by ShaoGongBra** [comment]
将podfile里面的hermes关掉就能运行起来
:hermes_enabled => false,

**Comment by zhiqingchen** [comment]
xcode 需要升级至 14.1

**Comment by zhiqingchen** [comment]
React Native 0.69 使用 Taro 3.6
===
旧版本从 3.5 升级到 3.6 或使用 Taro 3.5 初始化后，将 taro 相关版本升级到 3.6。需要做兼容。

## 方案一
使用 expo 47，将 ios 支持的最低版本升到 13.0
参考 https://github.com/NervJS/taro-native-shell/releases/tag/v0.69.7-2

1. 修改 Podfile
```
platform :ios, '13.0'
```

2. 修改 project.pbxproj
```
IPHONEOS_DEPLOYMENT_TARGET = 13.0;
```

## 方案二
使用 expo 46，将相关版本锁定在 0.69 对应版本
参考 https://github.com/NervJS/taro-native-shell/releases/tag/v0.69.7-3

1. 修改 package.json
```json
{
  "dependencies": {
    "expo": "~46.0.1",
    "expo-av": "~12.0.3",
    "expo-barcode-scanner": "~11.4.0",
    "expo-brightness": "~10.3.0",
    "expo-camera": "~12.3.0",
    "expo-file-system": "~14.1.0",
    "expo-image-picker": "~13.3.1",
    "expo-keep-awake": "~10.2.0",
    "expo-location": "~14.3.0",
    "expo-modules-core": "~0.11.3",
    "expo-sensors": "~11.4.0"
  },
  "resolutions": {
    "expo": "~46.0.1",
    "expo-av": "~12.0.3",
    "expo-barcode-scanner": "~11.4.0",
    "expo-brightness": "~10.3.0",
    "expo-camera": "~12.3.0",
    "expo-file-system": "~14.1.0",
    "expo-image-picker": "~13.3.1",
    "expo-keep-awake": "~10.2.0",
    "expo-location": "~14.3.0",
    "expo-modules-core": "~0.11.3",
    "expo-sensors": "~11.4.0"
  }
}
```

编译过程：https://github.com/NervJS/taro-native-shell/actions

React Native 0.68 使用 Taro 3.6
===
不进行验证，如需可参考 0.69 的方式

**Reply by Leslie-Wong-H** [reply]
求问：
Taro RN的uploadFile接口是不是存在域名证书验证问题？
总是会出现"TypeError: Network request failed"错误。

相关react native issue：
[HTTP Fetch fails with "TypeError: Network request failed" => Resolved](https://github.com/facebook/react-native/issues/32931)
[Fetch not working on Android 9 API 28](https://github.com/facebook/react-native/issues/24627)

Taro RN uploadFile实现源码：
https://github1s.com/NervJS/taro/blob/HEAD/packages/taro-rn/src/lib/file.ts#L60

**Reply by Leslie-Wong-H** [reply]
Taro Playground也没有看到uploadFile的接口用例，只有request和downloadFile的。

**Reply by zhiqingchen** [reply]
是不是使用了 http 的接口进行上传文件

**Reply by Leslie-Wong-H** [reply]
DNSPod的亚洲诚信证书

**Reply by zhiqingchen** [reply]
Taro 不处理证书相关内容，建议查阅 RN 相关issue
